api/REST-API.md

```diff
@@ -1,191 +1,596 @@
 
 Here are links to lessons that should be completed before this lesson:
 
 - [API/JSON Lesson](/api/apis-and-json.md)
 
-### Motivation
+## Motivation
+As you have seen in the API lesson, working with APIs is the daily bread of any full stack developer. An API opens a developer's code to the world, makes it easy for the world to interact with it. As a full stack developer, it is so rewarding to build your own API for your project. Believe me, it can be very rewarding to learn how to build one.
 
-REST is probably the most popular style of web API currently in use. It is not a framework or language, it's more like Object-Oriented Programming (OOP), in that it's a way to focus and design your system.
+## Content
 
-### Objectives
+[Part I: Lesson](#part-i:-lesson)
 
-- What is REST?
-- REST Features.
-- Guiding Principles of REST.
-- HTTP Methods
-- Advantages of REST
+- [What is an API?](#what-is-an-api)
+- [What is REST?](#what-is-rest)
+- [Guiding Principles of REST](#guiding-principles-of-rest)
+- [HTTP Methods](#http-methods)
+- [RESTful Routing](#restful-routing)
+- [PUT vs POST](#put-vs-post)
 
-### Lesson
+[PART II: Guided Practice](#part-ii-guided-practice)
 
-In a nutshell, REST is a style of API design that tries to be as close to standard HTTP as possible. Let's see what that means.
+- [Planning](#planning)
+- [Building the API](#building-the-api)
+- [Building the server](#building-the-server)
+- [Building the API](#building-the-api)
 
-#### Non-REST
+[PART III: Reference and Additional Readings](#part-iii:-reference-and-additional-readings)
 
-To understand REST, it's useful to first see the alternative. Network APIs existed before the Internet and before HTTP. Let's say you were designing an Eventonica web API. Your server is listening on port 7777. Any client that wants to use your API connects to this port and sends a JSON message using this HTTP request:
+- [Examples of real life APIs](#examples-of-real-life-apis)
+- [HTTP Status Codes](#http-status-codes)
+- [Additional Readings](#additional-readings)
+- [To learn more about the RESTful API constraints](#to-learn-more-about-the-restful-api-constraints)
 
-#### Requests
+[PART IV: Independent Practice](#part-iv:-independent-practice)
 
-##### `getEvent`
+---
 
-```
-POST / HTTP/1.1
+## PART I: Lesson
 
-{
-   "actionName": "getEvent",
-   "parameters": {
-      { "eventId": 123 }
-   }
-}
-```
+### What is an API? 
 
-Note that if you kind of squint, this almost looks like calling a function in JavaScript. In fact earlier APIs were designed as remote functions. And in your server code it will probably be handled by a function.
+As it stands, an API is an acronym for an **A**pplication **P**rogramming **I**nterface. It's a protocol; it's "the interface through which you access someone else’s code, or through which someone else's code accesses yours." ([APIs & JSON Lesson Slides](https://docs.google.com/presentation/d/1sD3nwQnhbe1wPnAWes0Nbt578tJacTtx0Yqy8XFp7w8/edit?usp=sharing)).
```

Comment:
Are these slides from the prerequisite lesson? Can you make it clear that they can get a refresher from the previous lesson? Just providing the link makes it seem like it’s *this outline’s* slideshow.

Reply:
Yes it is. Under the ## Prerequisites section I brought it in, but I added another hint here:

`As previously seen in the APIs & JSON Lesson Slides`

Reply:
see [1c62c05](https://github.com/Techtonica/curriculum/pull/1410/commits/1c62c05a5fef0b7f4e4a3eba5ae6459a5019094c)